### ISSUE TITLE
Use the updated HubConnection builder pattern in all the clients

### ISSUE THREAD

**Issue by mikaelm12** [aspnet/SignalR#3337]
In the Java client we introduced a new `HubConnection` builder pattern where `HubConnectionBuilder` was an abstract base class for builder implementations. The most common of which being the `HttpHubConnection` builder.
https://github.com/aspnet/SignalR/blob/c852bdcc332ffb998ec6a5b226e35d5e74d24009/clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnectionBuilder.java#L16-L21

We should use this pattern across the different clients.

**Comment by davidfowl** [comment]
There are some issues we need to discuss here, if the factory methods hang off of the builder itself, how do we handle different transports (http vs tcp for example).

**Comment by mikaelm12** [comment]
For example, in the http vs tcp case we could just use an overload 
`create(String url)`

**Comment by anurse** [comment]
This is pretty easy to do in TypeScript since the type system is super flexible. We need to think about how it works in .NET, but I think it's achievable.

For example, we can add `HubConnectionBuilder.Create()` that returns an `HttpHubConnectionBuilder` but what about the existing extension methods on `HubConnectionBuilder`, like `WithUrl` or `WithEndpoint`. If `HttpHubConnectionBuilder` inherits from `HubConnectionBuilder`, this code would work:

```csharp
var connection = HubConnectionBuilder.Create("http://mysite.com")
	.WithUrl("http://differentsite.com")
	.Build()
```

That would look weird and be confusing. I think a better approach is probably to have `HttpHubConnectionBuilder` *not* inherit from `HubConnectionBuilder`, but perhaps from some other base class or interface?

**Comment by davidfowl** [comment]
@mikaelm12 not quite. Those static methods cannot live on the base builder as they would be in different packages (potentially). I'd like to see a proposal using http and tcp as an example where each of those implementations lived in separate packages.

**Comment by davidfowl** [comment]
> I'd say we should remove those if we wanted to make HubConnectionBuilder the base class.

No, we can't make a breaking change like that to a core API that's used by *everyone*. The design needs to be one that doesn't break things. Pick a new base class (HubConnectionBuilderBase) name
